src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +144,162 @@ internal class DiagnosticComputer
+                //    request and add it to the tracked set of normal priority cancellation token sources.
+                //    This token source allows normal priority computeTasks to be cancelled when
+                //    a subsequent high priority diagnostic request is received.
+                using var cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                lock (s_gate)
+                {
+                    s_normalPriorityCancellationTokenSources = s_normalPriorityCancellationTokenSources.Add(cancellationTokenSource);
+                }
+
+                try
+                {
+                    // Step 3:
+                    //  - Execute the core compute task for diagnostic computation.
+                    return await GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo,
+                        cancellationTokenSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException ex) when (ex.CancellationToken == cancellationTokenSource.Token)
+                {
+                    // Step 4:
+                    //  - Attempt to re-execute this cancelled normal priority task by running the loop again.
+                    continue;
+                }
+                finally
+                {
+                    // Step 5:
+                    //  - Remove the 'cancellationTokenSource' for completed or cancelled task.
+                    //    For the case where the computeTask was cancelled, we will create a new
+                    //    'cancellationTokenSource' for the retry.
+                    lock (s_gate)
+                    {
+                        Debug.Assert(s_normalPriorityCancellationTokenSources.Contains(cancellationTokenSource));
+                        s_normalPriorityCancellationTokenSources = s_normalPriorityCancellationTokenSources.Remove(cancellationTokenSource);
+                    }
+                }
+            }
+
+            static async Task WaitForHighPriorityTasksAsync(CancellationToken cancellationToken)
+            {
+                // We loop continuously until we have an empty high priority task queue.
+                while (true)
+                {
+                    ImmutableHashSet<Task> highPriorityTasksToAwait;
+                    lock (s_gate)
+                    {
+                        highPriorityTasksToAwait = s_highPriorityComputeTasks;
+                    }
+
+                    if (highPriorityTasksToAwait.IsEmpty)
+                    {
+                        return;
+                    }
+
+                    // Wait for all the high priority tasks, ignoring all exceptions from it.
+                    await Task.WhenAll(highPriorityTasksToAwait).WithCancellation(cancellationToken).NoThrowAwaitable(false);
```

Comment:
actually, this is slightly wrong (my bad).  this means if we get canceled, this will no-throw, and we'll continue.

so likely best to add at least one cancellationTOken.ThrowIfRequested in this loop (maybe right after the `while`.

Reply:
Done, thanks